### PR TITLE
 Use `std::thread` instead of `OMP` for GPUs.

### PR DESCRIPTION
**Pull request by trivialfis**
Use std::thread in ExecuteIndexShards.

* Span for atomic write symbol.
* Use `ExecutePerDevice` instead of OpenMP loop.
* Use `SaveCudaContext` to make sure nothing can change master thread's device.

This decouples parameter `nthread` with `n_gpus`.  See  #4162.

I added some debugging utilities in device_helpers.cuh, please let me keep them ...

### CODE REVIEW

**Comment by trivialfis** [comment]
@hcho3 Can we upgrade the compiler at some point ... What the convention of supported compiler?

**Comment by hcho3** [comment]
@trivialfis Right now, we assume that C++11 is supported. Do you want something from C++14 or C++17?

**Comment by trivialfis** [comment]
@hcho3 No.  Just a compiler that has complete support for c++11.  For example you found `<regex>` is not available in g++-4.8.2,  and here the variadic arguments capturing is not working in 4.8.2.  See: https://stackoverflow.com/questions/20006574/workaround-for-variadic-lambda-capture

I might be able to workaround this later, but if we were to claim c++11 support, 4.8.2 might be too old.

**Comment by hcho3** [comment]
@trivialfis Got it, we should probably bump the version. Any suggestion?

**Comment by trivialfis** [comment]
@hcho3 I'm not sure about what's the status in commercial deployment.  Can you find an example for currently active distribution that uses 4.x?  If not, I would suggest sticking to one of those "popular stable distributions" like debian.

**Comment by hcho3** [comment]
@trivialfis 
> I would suggest sticking to one of those "popular stable distributions" like debian.

The reason for using CentOS 6 is to maximize the compatibility of the binary wheel in many Linux distributions. Otherwise, XGBoost wheel may depend on recent versions of GLIBC and other system libraries that other distributions may lack.

That said, we can certainly upgrade the GCC version. We just need to use later version of `devtoolset` package.

**Comment by trivialfis** [comment]
@hcho3 I see.  Just checked out CentOS, EOL is at the end of 2020 .....

Let's hold it and I will try to work around it for this PR.  Thanks.

**Comment by trivialfis** [comment]
One lesson I learned from this PR is, under no circumstance should one change the device in master thread.  I wish I can add a test for that.

**Comment by trivialfis** [comment]
@RAMitchell

File: src/common/device_helpers.cuh [hunk 1]
```diff
@@ -382,13 +433,17 @@ class BulkAllocator {
...
-    for (size_t i = 0; i < d_ptr_.size(); i++) {
-      if (!(d_ptr_[i] == nullptr)) {
-        safe_cuda(cudaSetDevice(device_idx_[i]));
-        safe_cuda(cudaFree(d_ptr_[i]));
-        d_ptr_[i] = nullptr;
+    SaveCudaContext {
+      [&]() {
+        for (size_t i = 0; i < d_ptr_.size(); i++) {
+          if (!(d_ptr_[i] == nullptr)) {
+            safe_cuda(cudaSetDevice(device_idx_[i]));
```
Comment (RAMitchell) [inline src/common/device_helpers.cuh]: I thought we established that setting the device was unnecessary for cudaFree?

File: src/common/device_helpers.cuh [hunk 2]
```diff
@@ -575,8 +631,8 @@ void SparseTransformLbs(int device_idx, dh::CubMemory *temp_memory,
   CHECK(num_tiles < std::numeric_limits<unsigned int>::max());
 
   temp_memory->LazyAllocate(sizeof(CoordinateT) * (num_tiles + 1));
-  CoordinateT *tmp_tile_coordinates =
-      reinterpret_cast<CoordinateT *>(temp_memory->d_temp_storage);
+  xgboost::common::Span<CoordinateT> tmp_tile_coordinates {
+    temp_memory->GetSpan<CoordinateT>(sizeof(CoordinateT) * (num_tiles + 1))};
```
Comment (RAMitchell) [inline src/common/device_helpers.cuh]: Changes relating to span should be in a different PR

File: src/linear/updater_gpu_coordinate.cu [hunk 3]
```diff
@@ -53,17 +53,13 @@ class DeviceShard {
...
-      };
-      auto column_begin =
-          std::lower_bound(col.cbegin(), col.cend(),
-                           xgboost::Entry(row_begin, 0.0f), cmp);
-      auto column_end =
-          std::lower_bound(col.cbegin(), col.cend(),
-                           xgboost::Entry(row_end, 0.0f), cmp);
-      column_segments.emplace_back(
-          std::make_pair(column_begin - col.cbegin(), column_end - col.cbegin()));
+      auto cmp = [](Entry e1, Entry e2) { return e1.index < e2.index; };
```
Comment (RAMitchell) [inline src/linear/updater_gpu_coordinate.cu]: These formatting changes make the PR confusing. It's not immediately obvious if the changes are only formatting and this file will be included in the diff even when nothing has changed relating to the commit message.

**Comment by hcho3** [comment]
@trivialfis For some reason, one of the GPU agents failed with error "Remote call on Linux GPU slave (i-0c8b6da83669c814e) failed". I restarted the agent.

**Comment by trivialfis** [comment]
@hcho3 Thanks for helping out.

@RAMitchell Done removing unrelated changes.

**Comment by RAMitchell** [comment]
Let me do a performance test then I will merge it if everything is okay.

**Comment by RAMitchell** [comment]
@trivialfis there is a performance regression here. Using the following command with 8x Tesla V100-SXM2-32GB:
```
python xgboost/tests/benchmark/benchmark_tree.py --rows 10000000 --params "{'n_gpus':8, 'debug_verbose':0}"
```
Time appears to have gone from around 20s to 50s. According to my recent profiling poor scaling on multi-GPU seems related to the way we are using threads.

Another way of achieving what you are trying to do while staying with omp would be to store the omp global number of threads (set according to nthreads parameter) to a temporary variable, change the number of threads to the number of GPUs just for ExecuteShards, then change it back afterwards.

**Comment by trivialfis** [comment]
@RAMitchell I wonder what happened ...  Let me see if I can reproduce the slow down.   BTW, `debug_verbose` is deprecated.

**Comment by trivialfis** [comment]
With #4454 it's possible to get rid of the global OpenMP parameter, I will close this one.
